restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java

```diff
@@ -489,12 +498,238 @@ public ReactiveDocumentService(
+                arrayData.add((Double) value);
+              }
+              return arrayData;
+            });
+  }
+
+  /**
+   * Retrieves the specified path in the document as an array, pops the last (i.e. highest-index)
+   * value, and returns both what the array will look like after the pop and the value that was
+   * popped. Note that this doesn't write the array back to the data store.
+   *
+   * @param db
+   * @param namespace
+   * @param collection
+   * @param documentId
+   * @param processedPath
+   * @param context
+   * @return a Pair containing the JSON representation of the array after popping the value and the
+   *     value itself
+   */
+  public Maybe<Pair<ArrayNode, Object>> getArrayAndValueAfterPop(
+      DocumentDB db,
+      String namespace,
+      String collection,
+      String documentId,
+      List<String> processedPath,
+      ExecutionContext context) {
+    return getDocument(db, namespace, collection, documentId, processedPath, null, context)
+        .map(
+            array -> {
+              JsonNode data = array.getData();
+              if (data == null || !data.isArray()) {
+                throw new ErrorCodeRuntimeException(
+                    ErrorCode.DOCS_API_SEARCH_ARRAY_PATH_INVALID,
+                    "The path provided to pop from has no array");
+              }
+              ArrayNode arrayData = (ArrayNode) data;
+              if (arrayData.size() == 0) {
+                throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_ARRAY_POP_OUT_OF_BOUNDS);
+              }
+              JsonNode value = arrayData.get(arrayData.size() - 1);
+              Object finalValue;
+              if (value.isDouble() || value.canConvertToLong()) {
+                double dblValue = value.asDouble();
+                long longValue = (long) dblValue;
+                if ((double) longValue == value.asDouble()) {
+                  finalValue = longValue;
+                } else {
+                  finalValue = dblValue;
+                }
+              } else if (value.isTextual()) {
+                finalValue = value.asText();
+              } else if (value.isBoolean()) {
+                finalValue = value.asBoolean();
+              } else if (value.isObject()) {
+                finalValue = objectMapper.treeToValue(value, Map.class);
+              } else if (value.isArray()) {
+                finalValue = objectMapper.treeToValue(value, List.class);
+              } else {
+                finalValue = null;
+              }
+              arrayData.remove(arrayData.size() - 1);
+              return Pair.with(arrayData, finalValue);
+            });
+  }
+
+  public Maybe<DocumentResponseWrapper<? extends JsonNode>> executeBuiltInFunction(
+      DocumentDB db,
+      String keyspace,
+      String collection,
+      String id,
+      ExecuteBuiltInFunction funcPayload,
+      List<PathSegment> path,
+      ExecutionContext context)
+      throws UnauthorizedException {
```

Comment:
where are we actually doing the authorization here?

Reply:
both `handlePush` and `handlePop` call `getDocument` which calls `getDocumentInternal` (and does the authentication piece)

Reply:
but getDocument is a read op? these are write ones?

Reply:
After reading, it calls `DocumentDB#deleteThenInsertBatch` which does write auth